posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,246 @@
+tease out the relationship between speed of the function and number of warnings
+thrown in a function with and without `cli::cli_warn()`. We will also check if
+the benchmark scales with the number of warnings in the function.
+
+Based on this description, we will need the following packages: `cli` for its warning function,
+[bench](https://cran.r-project.org/web/packages/bench/index.html) for measuring
+the run times, `purrr`, `tidyr`, and `dplyr` for running the functions and
+wrangling the results, and and `ggplot2` for visualizing the results.
+
+```{r}
+library(bench)
+library(cli)
+library(ggplot2)
+library(purrr)
+library(dplyr)
+library(tidyr)
+```
+
+Let's define a function that takes an argument `n` for the number of warnings
+to throw and `pkg` for the type of package to use.
+
+```{r}
+warn_Ntimes <- function(n, pkg) {
+  warning_msg <- "x must be an integer"
+  switch(pkg,
+    base = for (i in 1:n) {
+      warning(warning_msg)
+    },
+    cli = for (i in 1:n) {
+      cli_warn(warning_msg)
+    }
+  )
+}
+```
+
+Let's test our function to see if it works as expected.
+
+```{r}
+warn_Ntimes(3, "base")
+warn_Ntimes(3, "cli")
+```
+
+Now, we'll consider scenarios where a function throws 1, 5, 10, 15, 20, and
+100 warnings using base R and `cli`.
+
+```{r warning=FALSE}
+# Number of warnings to throw
+warnings <- c(1, 5, 10, 15, 20, 100)
+```
+
+Let's run benchmarks over the different scenarios and store results
+in a data.frame.
+
+```{r warning=FALSE,message=FALSE}
+bnmark_res <- bench::press(
+  warnings = c(1, 5, 10, 15, 20, 100),
+  bench::mark(
+    cli = warn_Ntimes(warnings, "cli"),
+    base = warn_Ntimes(warnings, "base")
+    )
+  )
+```
+
+## Results
+
+It's time to explore the results of the data generated. We'll start by making
+a boxplot of the median run times for the different scenarios.
+
+Drum roll please...
+
+```{r}
+autoplot(
+  bnmark_res,
+  type = "boxplot"
+  ) +
+  labs(
+    x = "Package",
+    caption = "Functions throwing various numbers of warnings as indicated in the facet label"
+  )
+```
+
+Finally, let's investigate the relative speeds of the two packages with
+respect to the number of warnings. We will do this by dividing the median
+benchmark of `cli::cli_warn()` by that of `base::warning()`.
+
+```{r}
+bnmark_relative_speeds <- bnmark_res |>
+  select(expression, median, warnings) |>
+  pivot_wider(
+    names_from = expression,
+    values_from = c(median)
+    ) |> 
+  mutate(
+    relative_speed = cli/base
+    )
+bnmark_relative_speeds
+```
+
+As we can see in the new column `relative_speed`, `cli` is consistently
+slower than base R and this seems not to
+scale linearly with the number of warnings thrown. Other benchmarking
+experiments
+with `cli` have revealed linear relationships between its speed and the number
+of times it invokes other functions [^3]. The developers of `cli` have also
+conducted benchmarks of the `ansi_` family of functions in `cli` in comparison
```

Comment:
Could add an asterisk (wildcard character) to `ansi_` (`ansi_*`) to show that this is a prefix for a family of functions.

Reply:
Thanks. I've made that change.